### PR TITLE
[Typeclasses] Warn when RHS of `:>` is not a class

### LINKED ISSUE #9014: In Class definitions, :> is silently ignored when the RHS is not a Class
When defining a class, using `:>` enables to declare “substructures”, as described in the [manual](https://coq.inria.fr/distrib/current/refman/addendum/type-classes.html?highlight=class#substructures). In the following example, `a_of_b` is declared as an instance, enabling to deduce an instance of `A` from an instance of `B`.

~~~
Variant T := mkT.
Class A := { p : Prop }.
Class B := { a_of_b :> A ; t_of_b :> T }.
~~~

However, when the right-hand-side of `:>` is not a class, as in `t_of_b` above, the “substructure” annotation is **silently** ignored.

In usual records, this annotation declares a coercion. Should this also happen with classes? At least, I suggest to emit a warning instead of only ignoring the annotation.

### PR DESCRIPTION
**Pull request by vbgl**
Fixes #9014.

Is there a better way of printing a `GlobRef.t`?

Should the projection be registered as a coercion instead?

NB: I also included in this PR an unrelated cleaning patch.

### CODE REVIEW

File: pretyping/typeclasses.ml
Comment (SkySkimmer) [inline pretyping/typeclasses.ml]: Use Nametab.shortest_qualid_of_global instead and print what it returns instead.

File: pretyping/typeclasses.mli
Comment (maximedenes) [inline pretyping/typeclasses.mli]: Why do you want to warn only sometimes?

Comment (vbgl) [reply pretyping/typeclasses.mli]: Because this function is called in other situations. E.g., when declaring section variables.

File: pretyping/typeclasses.ml
Comment (vbgl) [reply pretyping/typeclasses.ml]: Thus I have a similar question: how can I print a `types` (a.k.a. `Constr.t`)?

File: pretyping/typeclasses.mli
Comment (gares) [reply pretyping/typeclasses.mli]: I see. I think a comment (or the label name) should hint at when the warning is triggered

Comment (vbgl) [reply pretyping/typeclasses.mli]: Just added a comment.

File: pretyping/typeclasses.ml
Comment (SkySkimmer) [reply pretyping/typeclasses.ml]: Use Termops.Internal.print_constr_env

**Comment by vbgl** [comment]
@mattam82 Matthieu, would you like to comment on this PR? Thanks.

**Review by mattam82 — APPROVED** [review]
Looks good to me. The warning will surely help highlight the ambiguity with coercions to users.

**Comment by Zimmi48** [comment]
@mattam82 You're the only code owner so the natural assignee here.
That being said, @vbgl shouldn't this new warning get documented? (here: https://coq.inria.fr/refman/addendum/type-classes.html#substructures)

**Comment by mattam82** [comment]
@Zimmi48 right. The CI failure is spurious I suppose. I agree it needs a doc entry. And I don't think the projection should be declared as a coercion instead, to answer your initial question. We might want to move to an unambiguous syntax though (sorry for my own mistake pushing this confusion in the codebase initially :/).

**Comment by vbgl** [comment]
Just added documentation to the manual. Does it deserve a line in `CHANGES`?

**Comment by mattam82** [comment]
@vbgl I don't think so. I'll merge once CI passes.

**Comment by Zimmi48** [comment]
@mattam82 CI has passed.

**Review by Zimmi48 — APPROVED** [review]

**Comment by mattam82** [comment]
It might be good to turn this into an error in the following version

**Pull request merged**
